### PR TITLE
default to UTC when /etc/localtime is missing

### PR DESCRIPTION
**Pull request by esheppa**
This is one potential solution to #755. 

I've temporarily commented out some tests in src/offset/local/tz_info/timezone.rs, awaiting some discussion on where our "default to UTC" behavior should be (in unix.rs or in timezone.rs):
* If it belongs in `unix.rs` then the assumption of the test is invalidated as they will all panic
* If we move the logic to `timezone.rs` then the question is which errors should we recover from and which should we propogate (eg should filesystem errors cause a default to UTC as well?)
* When we lookup in the current offset we also use an `expect` - should this default to UTC instead of panicing:
```rust
                self.zone
                    .find_local_time_type(d.timestamp())
                    .expect("unable to select local time type")
                    .offset(),
```

### CODE REVIEW

**Comment by esheppa** [comment]
Ideally in `0.5` we could consider whether `now()` should be fallible which would then allow (and encourage by the type system) users to decide how to fallback

**Comment by djc** [comment]
Nice, thanks for working on this!

My inclination is to keep this scoped to Unix for now, which is where we broke compatibility with 0.4.20. I think for now we should stick to only defaulting to UTC if we get a NotFound error, and keep panicking for other types of errors. I'd like to hear about other types of failures before we tackle them.

Maybe we can also improve on the error message people see in this case, to make it explicit that it's about `/etc/localtime`, thereby making it easier to fix a problem if it comes up?

**Comment by djc** [comment]
Oh, and we could add a `try_now()`  already and expose it as fallible?

**Comment by djc** [comment]
I guess putting the logic in timezones.rs might work? I'm not sure, did you have a chance at how/when libc falls back?

**Comment by esheppa** [comment]
I tested this locally against date command which also seems to default to UTC in the case of no `/etc/localtime` symlink.

With regards to putting the logic in `timezone.rs` -  I've looked into this again and the caching logic also needs to handle fallback so it's nice to keep it in one file (albeit, there are probably further optimizations of the logic between `timezone.rs` and `unix.rs`.

However this may actually not be the right path - these cases where we are falling back to UTC due to a missing `/etc/localtime` make sense on docker, but not in android. We may need something like an `android.rs` and others, to handle platform specific logic/APIs. 

I'm also somewhat tempted by the suggestion of using `iana_time_zone` - however this would have to be in the form of a recommendation to use it with `chrono-tz` as it would need the compiled in time zone database. Potentially a long term solution is refactoring so that `chrono-tz` is merged into `chrono` and feature gated, then can be turned on for specific platforms local time implementations.

Alternatively to make this work in `chrono` for [android](https://android.googlesource.com/platform/ndk/+/android-4.2.2_r1.2/docs/system/libc/OVERVIEW.html):
* We have to first check `TZ` (which we do) and then check 'persist.sys.timezone' system property
* This only gives us the name of the selected timezone, which we then need to find in `/system/usr/share/zoneinfo`, then we can parse the correct one and continue as normal

**Comment by djc** [comment]
I think we should go the other way and rely on the system timezone database, I was thinking this morning about exposing other timezones in some way through the type system (where available).

You aren't quite explicit about why this solution doesn't work on Android, is that because we need to check this extra property? How would that work?

**Comment by djc** [comment]
Okay, after reviewing the code in iana-time-zone I can see that it would make sense to pull that in as a dependency.

**Comment by esheppa** [comment]
> I think we should go the other way and rely on the system timezone database, I was thinking this morning about exposing other timezones in some way through the type system (where available). 

This is an excellent idea IMO! As its better to have the automatically updated system tzdb (plus, a "static" feature could be possible for deployments to an environment lacking tzdb). 

> You aren't quite explicit about why this solution doesn't work on Android, is that because we need to check this extra property? How would that work?

We have to use the android SDK or equivalent to find the name of the currently chosen timezone

> Okay, after reviewing the code in iana-time-zone I can see that it would make sense to pull that in as a dependency. But not sure why that should be used with chrono-tz?

It only returns the name of the timezone, like "Asia/Kuala_Lumpur" or "America/Los_Angeles". This works if the system has the tzdb somewhere (eg Android has it, but in a different location), but `chrono-tz` is needed on systems that don't have a tzdb on the filesystem (potentially iOS)

**Comment by esheppa** [comment]
Ideally the fallback here would only be applied if all the following conditions are met:
* TZ environment variable is not set, or the file it points to is invalid
* Nothing at `/etc/localtime`
* iana-time-zone returns an error

**Comment by djc** [comment]
Yes, that sounds good.

I think we should prioritize shipping 0.4.21 that fixes this for systems that have a tzdb on the file system, we can take some more time to deal with iOS, I suppose (what did your "potentially" refer to?).

**Comment by esheppa** [comment]
One thing to note is that there are two main classes of failures we've seen currently:
* Alpine docker images -> no tzdb, no /etc/localtime -> solution: default to UTC (will use iana-time-zone and verify it returns an error before defaulting to UTC)
* Android -> has tzdb in different location, no /etc/localtime -> solution: use iana-time-zone to find zone name and then read the file from /system/usr/share/zoneinfo

Fixing these two gets us to a much better state than now, and then further work can be done for iOS etc (I'll raise separate issues for these)

**Comment by esheppa** [comment]
> For iOS I guess we could pull in the core_foundation crate and call CFTimeZone::system().seconds_from_gmt().
This is ok for `Local::now` but we also need `from_utc_datetime` and `from_local_datetime`

**Comment by djc** [comment]
I guess we can see about getting support for [`daylightSavingTimeOffset`](https://developer.apple.com/documentation/foundation/timezone/2293268-daylightsavingtimeoffset), but yeah, let's defer this until after 0.4.21.

File: Cargo.toml [hunk 1]
```diff
@@ -40,6 +40,7 @@ serde = { version = "1.0.99", default-features = false, optional = true }
 pure-rust-locales = { version = "0.5.2", optional = true }
 criterion = { version = "0.3", optional = true }
 rkyv = {version = "0.7", optional = true}
+iana-time-zone = { version = "0.1.41", optional = true }
```
Comment (djc) [inline Cargo.toml:43]: By the way, this requires an MSRV bump IIRC.

File: src/offset/local/unix.rs [hunk 2]
```diff
@@ -95,11 +96,27 @@ struct Cache {
     source: Source,
 }
 
+fn get_fallback_timezone() -> Option<TimeZone> {
+    let tz_name = try_opt!(iana_time_zone::get_timezone().ok());
+
+    #[cfg(target_os = "android")]
+    let tzdb_location = " /system/usr/share/zoneinfo";
```
Comment (djc) [inline src/offset/local/unix.rs:103]: Let's make these `const`?

File: Cargo.toml (hunk 1 above)
Comment (esheppa) [reply Cargo.toml:43]: hmm true. Potentially we could resolve this by vendoring parts but that is also not ideal

File: src/offset/local/unix.rs [hunk 3]
```diff
@@ -95,11 +96,27 @@ struct Cache {
     source: Source,
 }
 
+fn get_fallback_timezone() -> Option<TimeZone> {
+    let tz_name = try_opt!(iana_time_zone::get_timezone().ok());
```
Comment (djc) [inline src/offset/local/unix.rs:100]: I think we can actually just use `?` instead of `try_opt!()`?

File: Cargo.toml (hunk 1 above)
Comment (djc) [reply Cargo.toml:43]: Yeah, let's just go with 1.38 for now which is still fairly conservative. Actually might be a good middle step towards 1.48, giving downstream a chance to complain?

Comment (esheppa) [reply Cargo.toml:43]: sounds good to me! This gets us `TryFrom` as well which is very valuable IMO

**Comment by esheppa** [comment]
I've kept the MSRV change as a separate commit so it is clear when it happened

File: Cargo.toml (hunk 1 above)
Comment (djc) [reply Cargo.toml:43]: Agreed!

File: Cargo.toml [hunk 4]
```diff
@@ -40,12 +40,14 @@ serde = { version = "1.0.99", default-features = false, optional = true }
 pure-rust-locales = { version = "0.5.2", optional = true }
 criterion = { version = "0.3", optional = true }
 rkyv = {version = "0.7", optional = true}
-iana-time-zone = { version = "0.1.41", optional = true }
 
 [target.'cfg(all(target_arch = "wasm32", not(any(target_os = "emscripten", target_os = "wasi"))))'.dependencies]
 wasm-bindgen = { version = "0.2" }
 js-sys = { version = "0.3" } # contains FFI bindings for the JS Date API
 
+[target.'cfg(not(any(target_os = "emscripten", target_os = "wasi", target_os = "solaris")))'.dependencies]
```
Comment (djc) [inline Cargo.toml:48]: I think this belongs in the other commit?

File: src/offset/local/unix.rs [hunk 5]
```diff
@@ -99,18 +99,19 @@ struct Cache {
 #[cfg(target_os = "android")]
 const TZDB_LOCATION: &str = " /system/usr/share/zoneinfo";
 
+#[allow(dead_code)] // keeps the cfg simpler
```
Comment (djc) [inline src/offset/local/unix.rs:102]: The changes here also belong in the other commit, right?

File: src/offset/local/unix.rs [hunk 6]
```diff
@@ -95,11 +96,32 @@ struct Cache {
...
 }
 
+#[cfg(target_os = "android")]
+const TZDB_LOCATION: &str = " /system/usr/share/zoneinfo";
+
+#[cfg(not(target_os = "android"))]
+const TZDB_LOCATION: &str = "/usr/share/zoneinfo";
+
+#[cfg(any(target_os = "emscripten", target_os = "wasi", target_os = "solaris"))]
+fn get_fallback_timezone() -> Option<TimeZone> {
```
Comment (djc) [inline src/offset/local/unix.rs:107]: Let's strip the `get_()` prefix here in accordance with the API guidelines?

File: src/offset/local/tz_info/timezone.rs [hunk 7]
```diff
@@ -817,14 +817,16 @@ mod tests {
...
             } else {
-                let time_zone_local = TimeZone::local()?;
-                let time_zone_local_1 = TimeZone::from_posix_tz("localtime")?;
-                let time_zone_local_2 = TimeZone::from_posix_tz("/etc/localtime")?;
-                let time_zone_local_3 = TimeZone::from_posix_tz(":/etc/localtime")?;
-
-                assert_eq!(time_zone_local, time_zone_local_1);
-                assert_eq!(time_zone_local, time_zone_local_2);
-                assert_eq!(time_zone_local, time_zone_local_3);
+                // To be discussed - does this make sense in the context
```
Comment (djc) [inline src/offset/local/tz_info/timezone.rs:820]: What should we do with these tests now?

File: src/offset/local/unix.rs [hunk 8]
```diff
@@ -91,8 +97,9 @@ struct Cache {
 
 impl Default for Cache {
     fn default() -> Cache {
+        // default to UTC if no local timezone can be found
         Cache {
-            zone: TimeZone::local().expect("unable to parse localtime info"),
+            zone: TimeZone::local().unwrap_or_else(|_| TimeZone::utc()),
```
Comment (djc) [inline src/offset/local/unix.rs]: I think this can be written as `unwrap_or_else(TimeZone::utc)`?

**Review by djc — COMMENTED** [review]
Looking good!

File: Cargo.toml (hunk 4 above)
Comment (esheppa) [reply Cargo.toml:48]: perhaps the best solution is to squash the last two commits?

File: src/offset/local/tz_info/timezone.rs (hunk 7 above)
Comment (esheppa) [reply src/offset/local/tz_info/timezone.rs:820]: I think removing is the best option, they are just re-implementing the logic of `TimeZone::local()` anyway, and to make it work we'd also have to re-implement the UTC fallback here. The first case with TZ env var is useful as that verifies that the env var overrides other settings

Comment (djc) [reply src/offset/local/tz_info/timezone.rs:820]: Sounds good.

File: Cargo.toml (hunk 4 above)
Comment (djc) [reply Cargo.toml:48]: No, this is from the MSRV commit, which we'd like to keep separate. I'm fine with squashing the other two though.

File: src/offset/local/unix.rs (hunk 6 above)
Comment (djc) [reply src/offset/local/unix.rs]: Forgot to rename this one?

Comment (esheppa) [reply src/offset/local/unix.rs]: yep, sorted now

File: src/offset/local/unix.rs (hunk 8 above)
Comment (esheppa) [reply src/offset/local/unix.rs]: this should be sorted but the changes don't show up here

**Review by djc — APPROVED** [review]
Great!

**Comment by esheppa** [comment]
would be great to add android and iOS to the CI, will investigate in a seperate PR

**Pull request merged**